Implement trusted types integration with execCommand

lukewarlow opened the pull request:
#### 9fc59f12a6f2f4fc8289f4ec367d0944a3a2b4b1
<pre>
Implement trusted types integration with execCommand
<a href="https://bugs.webkit.org/show_bug.cgi?id=267690">https://bugs.webkit.org/show_bug.cgi?id=267690</a>

Reviewed by Youenn Fablet and Ryosuke Niwa.

Call trustedTypeCompliantString from within document execCommand when command is insertHTML.

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Document-execCommand-expected.txt:
* Source/WebCore/dom/Document+HTML.idl:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::execCommand):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/TrustedHTML.h:

Canonical link: <a href="https://commits.webkit.org/275667@main">https://commits.webkit.org/275667@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1247a06ac1f77fcf2e407f2e01c1c0331f8d1a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21426 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44802 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45007 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38525 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44715 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24653 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35127 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42982 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18359 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36522 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16069 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16011 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37567 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46475 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38602 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37896 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41801 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17224 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14189 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40420 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18843 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9496 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18905 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18488 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6863,13 +6864,35 @@ static Editor::Command command(Document* document, const String& commandName, bo
 }
 
-ExceptionOr<bool> Document::execCommand(const String& commandName, bool userInterface, const String& value)
+struct Visitor {
```
Remove this

youennf commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6895,13 +6896,27 @@ static Editor::Command command(Document* document, const String& commandName, bo
+                return String(str);
+            return trustedTypeCompliantString(TrustedType::TrustedHTML, *scriptExecutionContext(), str, "Document execCommand"_s);
+        },
+        [](RefPtr<TrustedHTML> trustedHtml) -> ExceptionOr<String> {
```
```suggestion
        [](const RefPtr<TrustedHTML>& trustedHtml) -> ExceptionOr<String> {
```

youennf commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6895,13 +6896,27 @@ static Editor::Command command(Document* document, const String& commandName, bo
     if (UNLIKELY(!isHTMLDocument() && !isXHTMLDocument()))
         return Exception { ExceptionCode::InvalidStateError, "execCommand is only supported on HTML documents."_s };
 
+    ExceptionOr<String> stringValueHolder = WTF::switchOn(value,
```
```suggestion
    auto stringValueHolder = WTF::switchOn(value,
```

youennf commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6895,13 +6896,27 @@ static Editor::Command command(Document* document, const String& commandName, bo
 }
 
-ExceptionOr<bool> Document::execCommand(const String& commandName, bool userInterface, const String& value)
+ExceptionOr<bool> Document::execCommand(const String& commandName, bool userInterface, std::variant<String, RefPtr<TrustedHTML>> value)
```
```suggestion
ExceptionOr<bool> Document::execCommand(const String& commandName, bool userInterface, const std::variant<String, RefPtr<TrustedHTML>>& value)
```
?

youennf commented on Source/WebCore/dom/TrustedHTML.h:
```diff
@@ -31,7 +31,7 @@
 namespace WebCore {
 
-class TrustedHTML : public ScriptWrappable, public RefCounted<TrustedHTML> {
+class WEBCORE_EXPORT TrustedHTML : public ScriptWrappable, public RefCounted<TrustedHTML> {
```
Why is it needed?

lukewarlow replied:
I believe this was needed for the build to succeed previously. I will just double check if it's still needed.

lukewarlow replied:
So it's needed else WebKitLegacy doesn't build, I think it's because Document::execCommand is exported.

lukewarlow committed:
Implement trusted types integration with execCommand
https://bugs.webkit.org/show_bug.cgi?id=267690

Reviewed by Youenn Fablet and Ryosuke Niwa.

Call trustedTypeCompliantString from within document execCommand when command is insertHTML.

* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Document-execCommand-expected.txt:
* Source/WebCore/dom/Document+HTML.idl:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::execCommand):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/TrustedHTML.h:

Canonical link: https://commits.webkit.org/275667@main
